[WebGPU+Swift] WebGPU.CommandEncoder.timestampWriteIndex does not handle partitioning correctly

mwyrzykowski opened the pull request:
#### 763e221cee89706e61ba12fa8e4afec85f949dec
<pre>
[WebGPU+Swift] WebGPU.CommandEncoder.timestampWriteIndex does not handle partitioning correctly
<a href="https://bugs.webkit.org/show_bug.cgi?id=297664">https://bugs.webkit.org/show_bug.cgi?id=297664</a>
<a href="https://rdar.apple.com/158779116">rdar://158779116</a>

Reviewed by Tadeu Zagallo.

Counter sample buffers get allocated at their max size (32KB) and then split into sub-sections to avoid per-process Metal framework limits.

This logic is missing from the Swift version

* Source/WebGPU/WebGPU/CommandEncoder.swift:
(WebGPU.timestampWriteIndex(_:defaultValue:offset:)):
(WebGPU.beginRenderPass(_:)):
(WebGPU.beginComputePass(_:)):

Canonical link: <a href="https://commits.webkit.org/300521@main">https://commits.webkit.org/300521@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ca8e2025e6ebafb685b6ea6a8af839d7f8857d9b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122795 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42507 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33197 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129421 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74899 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8c815680-a697-42f1-9e28-220ae0bee323) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43226 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51100 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93324 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61951 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/159bb682-9230-4e4c-95f5-34f151e8e999) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125747 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34465 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109918 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73965 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/04503e6c-2991-4b68-b61a-5eb31864aac9) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33447 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72913 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104165 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28287 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132147 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49740 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37867 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101834 "Passed tests") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/50236 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106134 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101709 "Found 3 new API test failures: WebKitGTK/TestInspector:/webkit/WebKitWebInspector/manual-attach-detach, TestWTF:CompletionHandlerDeathTest.MainThreadHandlerOnThreadAssertsDeathTest, TestWTF:CompletionHandlerDeathTest.ConstructionThreadHandlerOnThreadAssertsDeathTest (failure)") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25852 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25262 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/46510 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49597 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55350 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49064 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52416 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50747 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/474488a5b6ce543a4b8f89ec6c9850272b38464b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50260

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e8a636f30cfda0018e87a79b555008b7709d291d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50260

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d0c51eafdbcb29e1d1c67fed398696445f1c9376. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50260

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d43a71c39464171aa6130d000ecb428cf0d802d7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50260

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d7132b337b5f7fef71abd728d927d086d99220b8. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50260

mwyrzykowski committed:
[WebGPU+Swift] WebGPU.CommandEncoder.timestampWriteIndex does not handle partitioning correctly
https://bugs.webkit.org/show_bug.cgi?id=297664
rdar://158779116

Reviewed by Tadeu Zagallo.

Counter sample buffers get allocated at their max size (32KB) and then split into sub-sections to avoid per-process Metal framework limits.

This logic is missing from the Swift version

* Source/WebGPU/WebGPU/CommandEncoder.swift:
(WebGPU.timestampWriteIndex(_:defaultValue:offset:)):
(WebGPU.beginRenderPass(_:)):
(WebGPU.beginComputePass(_:)):

Canonical link: https://commits.webkit.org/300521@main
